06_Userspace/06_Resource_Management.md

```diff
@@ -0,0 +1,101 @@
+
+## Per Process Resource Table
+With an abstract resource now defined, we can extend our previous definition of a process to include a **resource table**:
+```
+typedef struct {
+size_t pid;
+status_t process_status;
+
+// Other fields
+
+resource_t* resource_table[MAX_RESOURCES];
+} process_t;
+```
+Now each process has a resource table that is a map of integers, called _handles_, to the kernel resource objects. A handle is simply an identifier returned by the kernel when opening a resource that is later used by the user to inform what resource the operation should be performed upon. This way, the resource structure is not exposed to userspace. Because of this, the same handle number in different processes can refer to different resources. For example, in Unix, handles `0`, `1`, and `2` refer to stdio for each process. 
+
+With this, we can also define a supporting function allowing the kernel to fetch a resource by handle:
+```
+resource_t* get_resource(process_t* proc, int handle) {
+
+  // Invalid handle
+  if (handle < 0 || handle >= MAX_HANDLES)
+      return NULL;
+  
+    return proc->table[handle];
+}
+```
+
+
+## Resource Lifecycle
+A resource follows a rather straightforward lifecycle, regardless of its type: 
+1. Firstly, a process acquires a handle by calling the `open_resource` system call.
+2. While the handle is valid, the process can perform operations such as `read_resource` or `write_resource`.
+3. Finally, when the process has finished using the resource, it calls `close_resource`, allowing the kernel to free any associated state.
+
+Typically, a process should `close()` a resource once it is done using it. However, that is not always the case, as processes may exit without cleaning up properly, and thus it is up to the kernel to ensure resources aren't leaked. This could look like a loop through the process's resource table, calling `close_resource(process, handle);` for each open resource and letting the resource-specific `close()` function handle the work.  
+
+
+## Generic API
+Now that we have a way of representing resource, we to need define how a process can interact with them. Generally, having a different syscall for each resource type can lead to lots of repeated code segments and make the kernel interface harder to maintain and extend. Instead the kernel can expose a minmal and uniform API that every resource supports. The generic interface for a resource consists of four primary functions: `open`, `read`, `write`, and `close` and by restricting all resources to this same interface we can reduce the complexity of both the kernel and userspace. To begin the implementation of this, our `resource_t` needs extending with a table of function pointers to support these operations. Each resource can then provide it's own implementation of thse four functions whilst the generic interface remains the same.
```

Comment:
representing resource -> representing resources

Reply:
This is a bit wordy after the first sentence, it sounds as if you're unsure of what you're writing. Have some confidence! You could try something more like:

> Having a different set of syscalls for each type of resource leads to a lot of duplicated code, and would be a massive API for programmers to use and maintain. Instead we chose a set of operations common to all resources, and have them operate on our abstract resources. This generic interfaces consists of four functions: `open`, `read`, `write` and `close`. To begin our implementation of this ... (rest is fine I think).